monitoring/nethealth.go

```diff
@@ -0,0 +1,321 @@
+	if err != nil {
+		return trace.Wrap(err, "failed to get nethealth metrics")
+	}
+
+	updated, err := c.updateStats(metrics)
+	if err != nil {
+		return trace.Wrap(err, "failed to update nethealth timeout stats")
+	}
+
+	return c.verifyNethealth(updated)
+}
+
+// getMetrics returns the network metrics from the local nethealth pod.
+func (c *nethealthChecker) getMetrics(ctx context.Context) ([]*dto.Metric, error) {
+	addr, err := c.getNethealthAddr()
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to get local nethealth address")
+	}
+
+	b, err := fetchMetrics(ctx, addr)
+	if err != nil {
+		log.WithError(err).Warnf("Failed to fetch metrics from %s", addr)
+		return nil, trace.Wrap(err, "failed to fetch metrics")
+	}
+
+	metrics, err := parseMetrics(bytes.NewReader(b))
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to parse metrics")
+	}
+
+	return metrics, nil
+}
+
+// updateStats updates the timeout data with new data points collected in
+// the provided metrics. Returns a list containing the updated keys.
+func (c *nethealthChecker) updateStats(metrics []*dto.Metric) (updated []string, err error) {
+	var errors []error
+	for _, metric := range metrics {
+		peerName, err := getPeerName(metric.GetLabel())
+		if err != nil {
+			errors = append(errors, trace.Wrap(err))
+			continue
+		}
+
+		series, err := c.getTimeoutSeries(peerName)
+		if err != nil {
+			errors = append(errors, trace.Wrap(err))
+			continue
+		}
+
+		// Keep only the last `seriesCapacity` number of data points.
+		if len(series) >= c.SeriesCapacity && len(series) > 0 {
+			series = series[1:]
+		}
+		series = append(series, int64(metric.GetCounter().GetValue()))
+
+		if err := c.setTimeoutSeries(peerName, series); err != nil {
+			errors = append(errors, trace.Wrap(err))
+		}
+
+		// Record updated nodes to be returned for use in later nethealth verification step.
+		updated = append(updated, peerName)
+	}
+	return updated, trace.NewAggregate(errors...)
+}
+
+// verifyNethealth verifies that the network communication is healthy for the
+// nodes specified by the provided list of names.
+func (c *nethealthChecker) verifyNethealth(names []string) error {
+	var errors []error
+	for _, name := range names {
+		series, err := c.getTimeoutSeries(name)
+		if err != nil {
+			errors = append(errors, trace.Wrap(err))
+			continue
+		}
+
+		if !c.isHealthy(series) {
+			errors = append(errors, trace.BadParameter("network communication failure with %s", name))
+		}
+	}
+	return trace.NewAggregate(errors...)
+}
+
+// isHealthy returns false if the number of timeouts increases at each data point.
+func (c *nethealthChecker) isHealthy(series []int64) bool {
+	// Checker has not collected enough data yet to check network health.
+	if len(series) < c.SeriesCapacity {
+		return true
+	}
+
+	// Increase in number of timeouts at each data point indicates network issue.
+	for i := 0; i < c.SeriesCapacity-1; i++ {
+		if series[i] >= series[i+1] {
```

Comment:
Trying to understand this algo - this kinda goes to my question about the data model above, what does `series[i]` represent? Number of timeouts during previous interval?

If so, won't this algorithm fail to detect the following scenario for example:

`5,6,5,6,5,6,5,6,..`

i.e. the number of timeouts decreases and the checker will think it's healthy. Unless I'm missing something (will prob be more clear when you document the data model).